driver/razermouse_driver.c

```diff
@@ -4666,6 +4686,139 @@ static void razer_mouse_init(struct razer_mouse_device *dev, struct usb_interfac
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_RECEIVER:
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_WIRED:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V2_X_HYPERSPEED:
+    case USB_DEVICE_ID_RAZER_VIPER_V2_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_VIPER_V2_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V3_PRO_WIRED:
+    case USB_DEVICE_ID_RAZER_DEATHADDER_V3_PRO_WIRELESS:
+    case USB_DEVICE_ID_RAZER_PRO_CLICK_MINI_RECEIVER:
+    case USB_DEVICE_ID_RAZER_VIPER_V3_HYPERSPEED:
+    case USB_DEVICE_ID_RAZER_VIPER_MINI_SE_WIRELESS:
+    case USB_DEVICE_ID_RAZER_VIPER_MINI_SE_WIRED:
+    case USB_DEVICE_ID_RAZER_HYPERPOLLING_WIRELESS_DONGLE:
+    case USB_DEVICE_ID_RAZER_NAGA_V2_HYPERSPEED_RECEIVER:
+        return 1;
+    default:
+        return 0;
+    }
+}
+
+int razer_scale_value(int value, int minIn, int maxIn, int minOut, int maxOut)
+{
+    // Ensure the input value is within the valid range.
+    if (value < minIn) value = minIn;
+    if (value > maxIn) value = maxIn;
+
+    return ((value - minIn) * (maxOut - minOut)) / (maxIn - minIn) + minOut;
+}
+
+static int razer_battery_get_property(struct power_supply *ps,
+                                      enum power_supply_property property, union power_supply_propval *val)
+{
+    struct razer_mouse_device *dev = power_supply_get_drvdata(ps);
+
+    int ret = 0;
+    switch (property) {
+    case POWER_SUPPLY_PROP_PRESENT:
+        val->intval = 1;
+        break;
+    case POWER_SUPPLY_PROP_SCOPE:
+        val->intval = POWER_SUPPLY_SCOPE_DEVICE;
+        break;
+    case POWER_SUPPLY_PROP_CAPACITY:
+        val->intval = razer_scale_value(razer_read_charge_level(dev), 0, 255, 0, 100);
+        break;
+    case POWER_SUPPLY_PROP_MODEL_NAME:
+        val->strval = razer_get_device_type(dev);
+        break;
+    case POWER_SUPPLY_PROP_SERIAL_NUMBER:
+        // this is the random serial number because we don't want upower to think the wired and wireless devices are the same
+        val->strval = dev->serial;
+        break;
+    case POWER_SUPPLY_PROP_ONLINE:
+        val->intval = 1;
+        break;
+    case POWER_SUPPLY_PROP_STATUS:
+        val->intval = razer_read_charge_status(dev)
+                      ? POWER_SUPPLY_STATUS_CHARGING : POWER_SUPPLY_STATUS_DISCHARGING;
+        break;
+    default:
+        ret = -EINVAL;
+        break;
+    }
+
+    return ret;
+}
+
+static enum power_supply_property razermouse_battery_props[] = {
+    POWER_SUPPLY_PROP_CAPACITY,
+    POWER_SUPPLY_PROP_MODEL_NAME,
+    POWER_SUPPLY_PROP_SERIAL_NUMBER,
+    POWER_SUPPLY_PROP_ONLINE,
+    POWER_SUPPLY_PROP_PRESENT,
+    POWER_SUPPLY_PROP_SCOPE,
+    POWER_SUPPLY_PROP_STATUS,
+};
+
+static DEFINE_IDA(razer_battery_device_id_allocator);
+
+static int razer_battery_init(struct hid_device *hdev, struct razer_mouse_device *dev)
+{
+    struct power_supply_config ps_config = {
+        .drv_data = dev
+    };
+
+    dev->battery_id = ida_alloc(&razer_battery_device_id_allocator, GFP_KERNEL);
```

Comment:
I didn't look much into this but looks like it's giving us a new unused integer that we use later then? Never heard of IDA before in the kernel

Reply:
it's id allocation and it's used so if there's multiple devices with batteries we don't create multiple power_supply devices with the same name

Reply:
ok, clear